great_expectations/datasource/fluent/interfaces.py

```diff
@@ -1243,3 +1257,123 @@ def _create_validator(self, *, result_format: ResultFormatUnion) -> V1Validator:
+            metric_configurations=[metric.config for metric in metrics],
+            runtime_configuration=None,
+        )
+        return self.metrics_calculator_result_to_metric_result(
+            requested_metric_names=requested_metric_names,
+            metrics_calculator_result=metrics_calculator_result,
+        )
+
+    @classmethod
+    def metrics_calculator_result_to_metric_result(
+        cls,
+        requested_metric_names: list[str],
+        metrics_calculator_result: tuple[_MetricsDict, _AbortedMetricsInfoDict],
+    ) -> MetricResult | list[MetricResult]:
+        """Converts the result of metrics_calculator.compute_metrics() into
+        a MetricResult or list[MetricResult]. Only the metrics that were
+        requested in the call to metrics_calculator.compute_metrics()
+        will be returned (dependency metrics are excluded)."""
+        # add another level of nesting to these metric dicts
+        # since the key we care about is nested inside the current key
+        # enables us to iterate through these dicts only once
+        raw_metrics = {k.metric_name: (k, v) for k, v in metrics_calculator_result[0].items()}
+        raw_metrics_errors = {
+            k.metric_name: (k, v) for k, v in metrics_calculator_result[1].items()
+        }
+        metric_results = []
+        metric_types: list[type[Metric]] = cls.get_all_metric_classes()
+        for metric_name in requested_metric_names:
+            if metric_name in raw_metrics:
+                MetricResultType = cls.get_metric_result_type_from_metric_name(
+                    metric_name=metric_name,
+                    metric_types=metric_types,
+                )
+                metric_results.append(
+                    MetricResultType(
+                        id=raw_metrics[metric_name][0],
+                        value=raw_metrics[metric_name][1],
+                    )
+                )
+            if metric_name in raw_metrics_errors:
+                metric_results.append(
+                    MetricErrorResult(
+                        id=raw_metrics_errors[metric_name][0],
+                        value=raw_metrics_errors[metric_name][1],
+                    )
+                )
+
+        if len(metric_results) == 1:
+            return metric_results[0]
+        return metric_results
+
+    @classmethod
+    def get_metric_result_type_from_metric_name(
+        cls,
+        metric_name: str,
+        metric_types: list[type[Metric]],
+    ) -> type[MetricResult]:
+        """Given a metric name and a list of Metric types gets the
+        MetricResult type for that Metric. Returns the base
+        MetricResult class if no MetricResult subclass is defined."""
+        for metric_type in metric_types:
+            if metric_type.name == metric_name:
+                return metric_type.get_metric_result_type()
+        return MetricResult
+
+    @classmethod
+    def get_all_metric_classes(cls) -> list[type[Metric]]:
```

Comment:
I'd prefer a registry over this type of introspection. Advantages of a registry:
* Contributors can register their metrics and don't have to add their metrics to this package. Note, we could get around this by using `Metric.__subclasses__()`. If we want to go this introspection route, I would prefer that to this. It also allows us to get rid of `METRICS_PACKAGE`.
* Registry lookup is faster since we don't need to do this traversal every time we call this method.

Reply:
Thanks for the tip. I added a runtime registry to `MetaMetric.__new__`.